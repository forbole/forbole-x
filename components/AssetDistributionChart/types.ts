export const sectoredByTypes = <const>['by assets', 'by validators'];

export type SectoredBy = typeof sectoredByTypes[number];
