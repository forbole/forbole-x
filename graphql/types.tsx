import { gql } from '@apollo/client'
import * as Apollo from '@apollo/client'

export type Maybe<T> = T | null
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] }
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> }
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> }
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string
  String: string
  Boolean: boolean
  Int: number
  Float: number
  /** The `Upload` scalar type represents a file upload. */
  Upload: any
}

// export type Query = {
//   __typename?: 'Query'
//   rates?: Maybe<Array<Maybe<ExchangeRate>>>
// }

// export type QueryRatesArgs = {
//   currency: Scalars['String']
// }

export type ExchangeRate = {
  __typename?: 'ExchangeRate'
  currency?: Maybe<Scalars['String']>
  rate?: Maybe<Scalars['String']>
  name?: Maybe<Scalars['String']>
}

export enum CacheControlScope {
  Public = 'PUBLIC',
  Private = 'PRIVATE',
}

export type GetAccountBalanceQueryVariables = Exact<{ [key: string]: never }>

export type GetAccountBalanceQuery = { __typename?: 'Query' } & {
  rates?: Maybe<Array<Maybe<{ __typename?: 'ExchangeRate' } & Pick<ExchangeRate, 'currency'>>>>
}

// export type GetCurrencyCadQueryVariables = Exact<{ [key: string]: never; }>;

// export type GetCurrencyCadQuery = (
//   { __typename?: 'Query' }
//   & {
//     rates?: Maybe<Array<Maybe<(
//       { __typename?: 'ExchangeRate' }
//       & Pick<ExchangeRate, 'currency'>
//     )>>>
//   }
// );

export const GetAccountBalanceQuery = gql`
  query AccountBalance($address: String, $height: bigint) {
    account(where: { address: { _eq: $address } }, limit: 1) {
      address
      available: account_balances(where: { height: { _eq: $height } }) {
        coins
      }
      delegations(where: { height: { _eq: $height } }) {
        amount
      }
      unbonding: unbonding_delegations(where: { height: { _eq: $height } }) {
        amount
      }
      rewards: delegation_rewards(
        where: { delegator_address: { _eq: $address }, height: { _eq: $height } }
      ) {
        amount
        withdraw_address
      }
    }
    validator: validator(where: { validator_info: { self_delegate_address: { _eq: $address } } }) {
      commissions: validator_commission_amounts(where: { height: { _eq: $height } }) {
        amount
      }
    }
  }
`

/**
 * __useGetAccountBalanceQuery__
 *
 * To run a query within a React component, call `useGetAccountBalanceQuery` and pass it any options that fit your needs.
 * When your component renders, `useGetAccountBalanceQuery` returns an object from Apollo Client that contains loading, error, and data properties
 * you can use to render your UI.
 *
 * @param baseOptions options that will be passed into the query, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options;
 *
 * @example
 * const { data, loading, error } = useGetAccountBalanceQuery({
 *   variables: {
 *   },
 * });
 */
export function useGetAccountBalanceQuery(
  baseOptions?: Apollo.QueryHookOptions<GetAccountBalanceQuery, GetAccountBalanceQueryVariables>
) {
  return Apollo.useQuery<GetAccountBalanceQuery, GetAccountBalanceQueryVariables>(
    GetAccountBalanceQuery,
    baseOptions
  )
}
export function useGetAccountBalanceLazyQuery(
  baseOptions?: Apollo.LazyQueryHookOptions<GetAccountBalanceQuery, GetAccountBalanceQueryVariables>
) {
  return Apollo.useLazyQuery<GetAccountBalanceQuery, GetAccountBalanceQueryVariables>(
    GetAccountBalanceQuery,
    baseOptions
  )
}
export type GetAccountBalanceQueryHookResult = ReturnType<typeof useGetAccountBalanceQuery>
export type GetAccountBalanceLazyQueryHookResult = ReturnType<typeof useGetAccountBalanceLazyQuery>
export type GetAccountBalanceQueryResult = Apollo.QueryResult<
  GetAccountBalanceQuery,
  GetAccountBalanceQueryVariables
>

// export const GetCurrencyCadDocument = gql`
//     query GetCurrencyCad {
//   rates(currency: "CAD") {
//     currency
//   }
// }
//     `;

// /**
//  * __useGetCurrencyCadQuery__
//  *
//  * To run a query within a React component, call `useGetCurrencyCadQuery` and pass it any options that fit your needs.
//  * When your component renders, `useGetCurrencyCadQuery` returns an object from Apollo Client that contains loading, error, and data properties
//  * you can use to render your UI.
//  *
//  * @param baseOptions options that will be passed into the query, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options;
//  *
//  * @example
//  * const { data, loading, error } = useGetCurrencyCadQuery({
//  *   variables: {
//  *   },
//  * });
//  */
// export function useGetCurrencyCadQuery(baseOptions?: Apollo.QueryHookOptions<GetCurrencyCadQuery, GetCurrencyCadQueryVariables>) {
//   return Apollo.useQuery<GetCurrencyCadQuery, GetCurrencyCadQueryVariables>(GetCurrencyCadDocument, baseOptions);
// }
// export function useGetCurrencyCadLazyQuery(baseOptions?: Apollo.LazyQueryHookOptions<GetCurrencyCadQuery, GetCurrencyCadQueryVariables>) {
//   return Apollo.useLazyQuery<GetCurrencyCadQuery, GetCurrencyCadQueryVariables>(GetCurrencyCadDocument, baseOptions);
// }
// export type GetCurrencyCadQueryHookResult = ReturnType<typeof useGetCurrencyCadQuery>;
// export type GetCurrencyCadLazyQueryHookResult = ReturnType<typeof useGetCurrencyCadLazyQuery>;
// export type GetCurrencyCadQueryResult = Apollo.QueryResult<GetCurrencyCadQuery, GetCurrencyCadQueryVariables>;