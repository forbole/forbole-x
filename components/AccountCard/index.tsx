import { Box, Card, Avatar, Typography, Button, useTheme } from '@material-ui/core'
import React from 'react'
import { LineChart, Line, YAxis } from 'recharts'
import UpIcon from '@material-ui/icons/ArrowDropUp'
import DownIcon from '@material-ui/icons/ArrowDropDown'
import useTranslation from 'next-translate/useTranslation'
import useStyles from './styles'
import { useSettingsContext } from '../../contexts/SettingsContext'
import { useAccountCardHook } from './hooks'

const AccountCard: React.FC = () => {
  const classes = useStyles()
  const theme = useTheme()
  const { lang } = useTranslation()
  const { currency } = useSettingsContext()
  // TODO: fetch data from backend
  const now = Date.now()
  // const balance = 104387.26485903
  const usdBalance = 626323.54
  const delta = new Array(24).fill(null).map(() => (Math.random() - 0.5) / 10)
  const data = []
  delta.forEach((d, i) => {
    data.unshift({
      time: now - i * 3600000,
      balance: i === 0 ? usdBalance : data[0].balance * (1 + d),
    })
  })
  const lastBalance = data[23].balance
  const firstBalance = data[0].balance
  const diff = Math.abs(lastBalance - firstBalance)
  const percentageChange = Math.round((10000 * diff) / firstBalance) / 100
  const increasing = lastBalance - firstBalance > 0

  // const denom = chainConfig.display.toUpperCase();

  const { accountInfo } = useAccountCardHook('desmos1qpm8wutycha3ncd0u3w9g42v89xnnfs6f9sg8d')
  // console.log('accountInfo', accountInfo)

  return (
    <Card className={classes.container}>
      <Box mb={7} display="flex" alignItems="center" justifyContent="space-between">
        <Box display="flex" alignItems="center">
          <Avatar
            alt="ATOM"
            src="https://research.binance.com/static/images/projects/cosmos-network/cosmoslogo.png"
          />
          <Typography className={classes.name} variant="h5">
            Account 12
          </Typography>
        </Box>
        <Button variant="outlined">Delegate</Button>
      </Box>
      <Box display="flex" alignItems="center" justifyContent="space-between">
        <Box>
          <LineChart width={theme.spacing(20)} height={theme.spacing(5)} data={data}>
            <YAxis domain={['dataMin', 'dataMax']} hide />
            <Line
              type="monotone"
              dataKey="balance"
              stroke={increasing ? theme.palette.success.main : theme.palette.error.main}
              dot={false}
              strokeWidth={2}
            />
          </LineChart>
          <Box display="flex" mt={1} alignItems="center">
            {increasing ? (
              <UpIcon htmlColor={theme.palette.success.main} />
            ) : (
              <DownIcon htmlColor={theme.palette.error.main} />
            )}
            <Box mr={2}>
              <Typography color="textSecondary">{percentageChange}% (24h)</Typography>
            </Box>
            <Typography>
              {new Intl.NumberFormat(lang, {
                style: 'currency',
                currency,
              }).format(diff)}{' '}
              {currency}
            </Typography>
          </Box>
        </Box>
        <Box display="flex" flexDirection="column" alignItems="flex-end">
          <Typography variant="h4" align="right">
            {/* {new Intl.NumberFormat(lang, {
              signDisplay: 'never',
            }).format(accountInfo.total.raw)} */}
            {accountInfo.total.display ? accountInfo.total.display : ''}
          </Typography>
          <Typography variant="h6" align="right">
            {new Intl.NumberFormat(lang, {
              style: 'currency',
              currency,
            }).format(usdBalance)}{' '}
            {currency}
          </Typography>
        </Box>
      </Box>
    </Card>
  )
}

export default AccountCard
