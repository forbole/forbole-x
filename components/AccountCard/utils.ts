import { formatDenom } from '../../utils'
import { AccountBalance } from '../../models'
import { chainConfig } from '../../config/chain_config'

export const formatData = (data: AccountBalance) => {
  return {
    address: data.address,
    total: formatDenom(chainConfig.display, data.total),
  }
}
