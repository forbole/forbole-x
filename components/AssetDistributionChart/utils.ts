import numeral from 'numeral'
import { AssetInfo } from '../../models'

export const formatData = (data: AssetInfo, crypto: string) => {
  return {
    address: data.address,
    total: data.total,
    delegate: data.delegate,
    crypto,
    delegationDetail: data.delegationDetail.map((delegation) => {
      return {
        validatorAddress: delegation.validatorAddress,
        validatorMoniker: delegation.validatorMoniker,
        amount: delegation.amount,
      }
    }),
  }
}

interface Summery {
  totalAmount: number
  delegation: {
    validatorMoniker: string
    amount: number
    value?: string
  }[]
  coinDetail: {
    crypto: string
    amount: number
    value?: string
  }[]
}

interface FormatData {
  address: string
  total: number
  delegate: number
  crypto: string
  delegationDetail: {
    validatorAddress: string
    validatorMoniker: string
    amount: number
  }[]
}

export const summarizedData = (dataList: FormatData[]) => {
  const summery: Summery = {
    totalAmount: 0,
    delegation: [],
    coinDetail: [],
  }

  if (dataList.length !== 0) {
    dataList.forEach((data) => {
      summery.totalAmount += data.delegate
      data.delegationDetail.forEach((validator) => {
        const index = summery.delegation.findIndex(
          (x) => x.validatorMoniker === validator.validatorMoniker
        )
        if (index !== -1) {
          summery.delegation[index].amount += validator.amount
        } else {
          summery.delegation.push({
            validatorMoniker: validator.validatorMoniker,
            amount: validator.amount,
          })
        }
      })

      const indexCoin = summery.coinDetail.findIndex((i) => data.crypto === i.crypto)
      console.log('indexCoin', indexCoin)
      if (indexCoin !== -1) {
        summery.coinDetail[indexCoin].amount += data.total
      } else {
        summery.coinDetail.push({
          crypto: data.crypto,
          amount: data.total,
        })
      }

    })
    summery.delegation.forEach((x) => {
      // eslint-disable-next-line no-param-reassign
      x.value = numeral(x.amount / summery.totalAmount).format('0.00%')
    })

    summery.coinDetail.forEach((x) => {
      // eslint-disable-next-line no-param-reassign
      x.value = numeral(x.amount / summery.totalAmount).format('0.00%')
    })

    return summery
  }
  return {
    totalAmount: 0,
    delegation: [
      {
        validatorMoniker: '',
        amount: 0,
        value: '0.00%',
      },
    ],
    coinDetail: [
      {
        crypto: '',
        amount: 0,
        value: '0.00%',
      },
    ],
  }
}
