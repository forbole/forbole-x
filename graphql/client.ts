import { ApolloClient, InMemoryCache, ApolloLink } from '@apollo/client';
import { concatPagination } from '@apollo/client/utilities';
import { useMemo } from 'react';
import { MultiAPILink } from '@habx/apollo-multi-endpoint-link';
import { createHttpLink } from 'apollo-link-http';
import cryptocurrencies from '../misc/cryptocurrencies';

const defaultOptions: any = {
  watchQuery: {
    // fetchPolicy: 'no-cache',
    // errorPolicy: 'ignore',
  },
  query: {
    // fetchPolicy: 'no-cache',
    // errorPolicy: 'all',
  },
};

let apolloClient;
const endpoints = {};

Object.values(cryptocurrencies).forEach(c => {
  endpoints[`${c.name}bdjuno`] = c.graphqlHttpUrl;
  endpoints[`${c.name}djuno`] = c.djunoUrl;
});

function createApolloClient() {
  const client = new ApolloClient({
    ssrMode: typeof window === 'undefined',
    link: ApolloLink.from([
      new MultiAPILink({
        endpoints,
        createHttpLink: () => (createHttpLink as any)(),
      }),
    ]),
    cache: new InMemoryCache({
      typePolicies: {
        Query: {
          fields: {
            allPosts: concatPagination(),
          },
        },
      },
    }),
  });

  client.defaultOptions = defaultOptions;

  return client;
}

export function initializeApollo(initialState = null) {
  // eslint-disable-next-line
  const _apolloClient = apolloClient ?? createApolloClient()

  // If your page has Next.js data fetching methods that use Apollo Client, the initial state
  // gets hydrated here
  if (initialState) {
    // Get existing cache, loaded during client side data fetching
    const existingCache = _apolloClient.extract();
    // Restore the cache using the data passed from getStaticProps/getServerSideProps
    // combined with the existing cached data
    _apolloClient.cache.restore({
      ...existingCache,
      ...initialState,
    });
  }
  // For SSG and SSR always create a new Apollo Client
  if (typeof window === 'undefined') return _apolloClient;
  // Create the Apollo Client once in the client
  if (!apolloClient) apolloClient = _apolloClient;

  return _apolloClient;
}

export function useApollo(initialState) {
  const store = useMemo(() => initializeApollo(initialState), [initialState]);
  return store;
}
