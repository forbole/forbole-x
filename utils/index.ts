import { formatDenom } from './format_denom'

export { formatDenom }
