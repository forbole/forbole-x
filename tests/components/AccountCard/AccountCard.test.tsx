import React from 'react'
import renderer from 'react-test-renderer'
import { ApolloProvider } from '@apollo/client'
import { createMockClient, MockApolloClient } from 'mock-apollo-client'
import { GetAccountBalanceQuery } from '../../../graphql/types'
import { wait } from '../../utils'
import AccountCard from '../../../components/AccountCard'

let component
let mockClient: MockApolloClient

const accountBalanceMockData = jest.fn().mockResolvedValue({
  data: {
    account: [
      {
        address: 'desmos1qpm8wutycha3ncd0u3w9g42v89xnnfs6f9sg8d',
        available: [
          {
            coins: [
              {
                denom: 'udaric',
                amount: '9000155',
              },
            ],
          },
        ],
        delegations: [
          {
            amount: {
              denom: 'udaric',
              amount: '990000',
            },
          },
        ],
        unbonding: [],
        rewards: [
          {
            amount: [],
            withdraw_address: 'desmos1qpm8wutycha3ncd0u3w9g42v89xnnfs6f9sg8d',
          },
        ],
      },
    ],
    validator: [],
  },
})

const mockAccount = {
  walletId: 'f149b58827224d515a0025f991b240749b4e7927f47c55c7aee0da2d953e9312',
  address: 'desmos1qpm8wutycha3ncd0u3w9g42v89xnnfs6f9sg8d',
  createdAt: 1614826912972,
  crypto: 'DSM',
  fav: false,
  index: 0,
  name: 'DSM',
  displayName: '',
  rpDisplayName: '',
  id: '',
}

describe('AccountCard', () => {
  beforeEach(async () => {
    mockClient = createMockClient()
    mockClient.setRequestHandler(GetAccountBalanceQuery, accountBalanceMockData)

    const mockFormatExpression = {
      format: (value) => value.toString(),
    }
    let mockNumberFormatFunction = Intl.NumberFormat = jest.fn()
      .mockImplementation(() => mockFormatExpression);

    renderer.act(() => {
      component = renderer.create(
        <ApolloProvider client={mockClient}>
          <AccountCard account={mockAccount} />
        </ApolloProvider>
      )
    })
    await wait()
  })

  it('it renders', async () => {
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    expect(accountBalanceMockData).toHaveBeenCalled()
  })
})

afterEach(() => {
  jest.clearAllMocks()
})
