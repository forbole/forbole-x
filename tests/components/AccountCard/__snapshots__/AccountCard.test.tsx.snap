// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountCard it renders 1`] = `
<div
  className="MuiPaper-root MuiCard-root HookGlobalStyles-container-1 MuiPaper-elevation1 MuiPaper-rounded"
>
  <div
    className="MuiBox-root MuiBox-root-2"
  >
    <div
      className="MuiBox-root MuiBox-root-3"
    >
      <div
        className="MuiAvatar-root MuiAvatar-circle"
      >
        <img
          alt="DSM"
          className="MuiAvatar-img"
          src="/static/images/cryptocurrencies/dsm.png"
        />
      </div>
      <div
        className="MuiBox-root MuiBox-root-4"
      >
        <h5
          className="MuiTypography-root MuiTypography-h5"
        >
          DSM
        </h5>
        <button
          className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiLink-button MuiTypography-body2 MuiTypography-colorTextSecondary"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
        >
          desmos1qpm8wutycha3ncd0u3w9g42v89xnnfs6f9sg8d
          <div
            className="MuiBox-root MuiBox-root-5"
          >
            <svg
              data-file-name="SvgIcon_copy"
              fill="#e0e0e0"
              height={16}
              width={16}
            />
          </div>
        </button>
      </div>
    </div>
    <button
      className="MuiButtonBase-root MuiIconButton-root"
      disabled={false}
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          data-file-name="SvgIcon_more"
          fill="#e0e0e0"
          height={16}
          width={16}
        />
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
  </div>
  <div
    className="MuiBox-root MuiBox-root-6"
  >
    <div
      className="MuiBox-root MuiBox-root-7"
    >
      <h4
        className="MuiTypography-root MuiTypography-h4"
      >
        0
         
        DSM
      </h4>
      <h6
        className="MuiTypography-root MuiTypography-h6"
      >
        626323.54
         
        USD
      </h6>
    </div>
    <button
      className="MuiButtonBase-root MuiIconButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          data-file-name="SvgIcon_star"
          fill="#e0e0e0"
          height={16}
          width={16}
        />
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
